tests/apollo/test_skvbc_preexecution.py

```diff
@@ -191,3 +192,33 @@ async def test_view_change(self, bft_network):
                 break
 
         self.assertEqual(new_last_block, last_block)
+
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    async def test_parallel_tx_after_f_nonprimary_crash(self, bft_network):
+        '''
+        Crash f nonprimary replicas and submit X parallel write submissions.
+        Block processing of the network should be unaffected with f-count interruption.
+        Final block length should match submitted transactions count exactly.
+        '''
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+        await trio.sleep(SKVBC_INIT_GRACE_TIME)
+        
+        read_client = bft_network.random_client()
+        submit_clients = bft_network.random_clients(MAX_CONCURRENCY)
+        num_of_requests = 10 * len(submit_clients) # each client will send 10 tx
+        nonprimaries = bft_network.all_replicas(without={0}) # primary index is 0
+        crash_targets = random.sample(nonprimaries, bft_network.config.f) # pick random f to crash
+        bft_network.stop_replicas(crash_targets) # crash chosen nonprimary replicas
+
+        await self.run_concurrent_pre_execution_requests(skvbc, submit_clients, num_of_requests, write_weight=1.0)
+        final_block_count = skvbc.parse_reply(await read_client.read(skvbc.get_last_block_req()))
+
+        print("")
+        print(f"Randomly picked replica indexes {crash_targets} (nonprimary) to be stopped.")
+        print(f"Total of {num_of_requests} write pre-exec tx, "
+              f"concurrently submitted through {len(submit_clients)} clients.")
+        print(f"Finished at block {final_block_count} (expected {num_of_requests}).")
+        self.assertEqual(final_block_count, num_of_requests)
```

Comment:
Due to batching, the best we can assert at this point is the following:
`self.assertTrue(final_block_count >= num_of_requests / len(submit_clients))`.

Can you please check if the test is stable with this condition?

Thanks!

Reply:
I did a 100 test run with following:
```bash
#!/bin/bash
success_count = 0; fail_count = 0
for i in {1..100}; do
    BUILD_ROCKSDB_STORAGE=ON STORAGE_TYPE=v1direct \
    python3 -m unittest test_skvbc_preexecution.SkvbcPreExecutionTest.test_parallel_tx_after_f_nonprimary_crash
    if [ $? -eq 0 ]; then ((success_count++)); else ((fail_count++)); fi
done
echo "test $success_count success / $fail_count failed"
```
I got, test 94 success / 6 failed.
I think 6 failures are from `TooSlow` errors.

Reply:
Seems to work well with `self.assertTrue(final_block_count >= num_of_requests / len(submit_clients))`

Still trying to investigate where `TooSlow` errors are coming from.

Reply:
`TooSlow` with following output from primary:
```
4|13-06-2020 23:09:24.747|INFO |skvbctest.replica|message-processing|0|6677708463537127424|2|slow|bool InternalCommandsHandler::executeWriteCommand(uint32_t, const char*, uint64_t, uint8_t, size_t, char*, uint32_t&)|ConditionalWrite message handled; writesCounter=9 currBlock=9
4|13-06-2020 23:09:24.747|INFO |concord|message-processing||6677708463541321728|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463541321728 from clientId=12, reqTimeoutMilli=5000, senderId=12
2|13-06-2020 23:09:24.747|INFO |concord|message-processing||6677708463541321728|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463541321728 from clientId=12, reqTimeoutMilli=5000, senderId=12
2|13-06-2020 23:09:24.747|INFO |concord|message-processing||6677708463541321728|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463541321728 has already been executed - let replica to decide how to proceed further
2|13-06-2020 23:09:24.747|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=33, reqTimeoutMilli=5000, senderId=33
4|13-06-2020 23:09:24.747|INFO |concord|message-processing||6677708463541321728|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463541321728 has already been executed - let replica to decide how to proceed further
4|13-06-2020 23:09:24.748|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=33, reqTimeoutMilli=5000, senderId=33
2|13-06-2020 23:09:24.748|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
4|13-06-2020 23:09:24.748|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:24.832|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:24.832|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:25.085|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:25.085|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:25.338|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:25.338|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:25.589|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:25.590|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:25.841|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:25.842|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:26.093|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:26.093|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:26.344|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:26.345|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:26.597|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:26.597|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:26.848|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:26.848|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:27.099|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:27.099|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:27.351|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:27.352|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:27.605|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:27.605|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:27.857|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:27.858|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:28.110|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:28.110|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
0|13-06-2020 23:09:28.363|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|Going to process ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 from clientId=23, reqTimeoutMilli=5000, senderId=23
0|13-06-2020 23:09:28.363|INFO |concord|message-processing||6677708463532933120|||void preprocessor::PreProcessor::onMessage(T*) [with T = preprocessor::ClientPreProcessRequestMsg]|ClientPreProcessRequestMsg reqSeqNum=6677708463532933120 has already been executed - let replica to decide how to proceed further
TooSlowError thrown from client_id 23 for seq_num 6677708463532933120
E
```

Reply:
This output seems to imply the client request has already been processed, but the client hasn't received the result and is probably retrying the request.

Can you please ask Yulia for help on this one? Thanks!